### PR TITLE
Feature/enable build publish workflow

### PR DESCRIPTION
**Pull request by Tatsinnit**
Opening this for wider eyes and as PR for enabling `GHCR publishing` for the repo, please note this can also be a blue print PR for other repos Matt Stam + other projects et. al. are working for the publishing of the `ACR` Images, only thing which will change is the registry and mechanism of publishing.

Here is a quick detail of actions:

* Workflow is on `trigger` only, (So admin for the repo can do that)
* Then, workflow checkout repo, then it used `ChangeLog` action for checking out the version. https://github.com/mindsers/changelog-reader-action
* After that, the workflow triggers the `GHCR` publishing mechanism.

Thanks heaps,

### CODE REVIEW

**Review by gambtho — APPROVED** [review]

File: .github/workflows/build-and-publish.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        env:
+            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} 
+        with:
+          tag_name : ${{ steps.changelog_reader.outputs.current-version}}
+          release_name: ${{ steps.changelog_reader.outputs.current-version}}
+          body: Publish ${{ steps.changelog_reader.outputs.current-version}}
+      # Lowercase my github ownername.
+      - name: Set Environment Variables
+        run: |
+          OWNER="$(echo "${{ github.repository_owner }}" | tr '[:upper:]' '[:lower:]')"
```
Comment (mattstam) [inline .github/workflows/build-and-publish.yml:55]: Where is `repository_owner` property set in a given repo? 

Is this something we instead want to allow for multiple people, such as everyone under a certain team (ie. https://github.com/orgs/Azure/teams/agentbakerwrite/members)?

**Review by mattstam — APPROVED** [review]
Had a few questions but LGTM, thanks for this!

File: .github/workflows/build-and-publish.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        id: changelog_reader
+        uses: ./
+      - name: Display output
+        run: |
+          echo "Version: ${{ steps.changelog_reader.outputs.version }}"
+      - name: Create a Release
+        id: create_release
+        uses: actions/create-release@v1
+        env:
+            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (Tatsinnit) [reply .github/workflows/build-and-publish.yml:47]: Indeed, thank you for asking and I will add it for clarity, Yes, at the Org level this is default variable passed, I think as Admin we only need to limit the access of folkid who can do this. (No var setting is in need.) :)

File: .github/workflows/build-and-publish.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the main branch
+on: [workflow_dispatch]
+
+env:
+  GO_VERSION: '1.16.1'
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
```
[the remark this answers is not in the document]
Comment (Tatsinnit) [reply .github/workflows/build-and-publish.yml:12]: 💡 Thank you for this, so, in reply I would say `No` this is the workflow for `Build and release` the `ci-pipeline` triggers for every pull-request` and `push` to master. 

* We don't want release to trigger every push, `Release + Build` is an `Exclusive` separate workflow hence its is treated separately.

There is few good material about it, but one good pipeline to see is vscode pipelines where its clearly separated and easy to handle, and `Access Right` to these `Workflow` is `Permission Bound`

File: .github/workflows/build-and-publish.yml (hunk 1 above)
Comment (Tatsinnit) [reply .github/workflows/build-and-publish.yml:55]: Indeed, `repository_owner ` this is default set at github level. Thank you for asking.

* For more details this is a **Good** read: https://docs.github.com/en/actions/learn-github-actions/environment-variables 

* I would say as long as this list is safe and actual contributor to the project then there is no hard, we trust these folks, but if that is not the case, best is to remove those aliases from the list of access.

* Idea is no - I would say - No one will go and triggers release for fun :) especially it is bound to the `ChangeLog`

**Pull request merged**